### PR TITLE
Add sensor support

### PR DESCRIPTION
**Pull request by herbrechtsmeier**
Add support for Hue sensors. Move common parts from HueLight class to a new HueThing class and add a new HueSensor class.

**Questions**
- Why are all functions are virtual?

**ToDo**
- Move unit tests of HueThing
- Add unit tests for HueSensor

### CODE REVIEW

**Comment by Jojo-1000** [comment]
Great work! Our goal is to support most Hue API functions in the future. 

I am currently working on adding group support (branch group-api), which contains some changes that also affect HueLight.

> * Why are all functions are virtual?

They are virtual for the test mock classes.

File: src/Hue.cpp [hunk 1]
```diff
@@ -445,7 +489,7 @@ void Hue::refreshState()
         return;
     }
     nlohmann::json answer = commands.GETRequest("", nlohmann::json::object(), CURRENT_FILE_INFO);
-    if (answer.is_object() && answer.count("lights"))
+    if (answer.is_object() && (answer.count("lights") || answer.count("sensors")))
```
Comment (enwi) [inline src/Hue.cpp:492]: Wouldn't an explicit `bool Hue::sensorExists(int id)` be better? At least to reduce confusion?

File: src/Hue.cpp [hunk 2]
```diff
@@ -246,6 +246,34 @@ HueLight& Hue::getLight(int id)
...
+    refreshState();
+    if (!state["sensors"].count(std::to_string(id)))
+    {
+        std::cerr << "Error in Hue getSensor(): sensor with id " << id << " is not valid\n";
+        throw HueException(CURRENT_FILE_INFO, "Sensor id is not valid");
+    }
+    // std::cout << state["sensors"][std::to_string(id)] << std::endl;
+    std::string type = state["sensors"][std::to_string(id)]["modelid"];
+    // std::cout << type << std::endl;
+    if (type == "RWL021" || type == "PHDL00" || type == "PHWA01")
```
Comment (enwi) [inline src/Hue.cpp:266]: This should probably be added to HueDeviceTypes.cpp

**Comment by sieren** [comment]
That's pretty cool!
Not sure how deep you'd plan on going with this, but I suppose it's worth tkaing over the support devices from the deCONZ project (which covers pretty much all Zigbee smart home appliances that work with Hue) ( https://github.com/dresden-elektronik/deconz-rest-plugin/blob/master/de_web_plugin.cpp L106+)

I might look into that at some point

**Comment by herbrechtsmeier** [comment]
@Jojo-1000 Isn't it strange to adapt the source code to a tests?

File: src/Hue.cpp (hunk 2 above)
Comment (herbrechtsmeier) [reply src/Hue.cpp:266]: Should we really manage lights and sensors in one class? Does the types are shared between lights and sensors?

**Comment by herbrechtsmeier** [comment]
@sieren Are you sure that all this devices are supported by the Hue bridge? Doesn't the hue bridge only support ZigBee Light Link?

**Comment by herbrechtsmeier** [comment]
Should I rename `HueThing` into `HueDevice` because this term is already used?

**Comment by sieren** [comment]
@herbrechtsmeier Youre right, I think thats where they diverge a little. DeCONZ certainly supports a lot more devices than the Hue bridge (on a Zigbee level), but the way they are exposed through the REST API is consistent with the Hue Bridge.

I'm still interested in adding those to Hueplusplus since deCONZ is a much cheaper and way better zigbee hub than the Philipps/Signify Bridge.

**Comment by enwi** [comment]
> Should I rename `HueThing` into `HueDevice` because this term is already used?

Regarding my requested change this would make sense and then you could also add the 'Device' types to HueDeviceTypes.cpp

**Comment by herbrechtsmeier** [comment]
@sieren The deCONZ REST API also supports Websocket for live updates.

**Comment by herbrechtsmeier** [comment]
> > Should I rename `HueThing` into `HueDevice` because this term is already used?
> 
> Regarding my requested change this would make sense and then you could also add the 'Device' types to HueDeviceTypes.cpp

Should I add a `MakeHueSensor::operator()`?

I wonder if we should add a `HueSensors` class instead and rework the data holding. At the moment the `light` resources are buffered in the `Hue' and `HueLight` class and differently updated. I would move the data holding into the `HueSensor` class. The `HueSensors` class can forward its information to the `HueSensor` class and the `Hue` class to the `HueSensors` class. Especially for the sensors it makes sense to update all sensors together because its maybe faster.

**Comment by enwi** [comment]
@herbrechtsmeier @Jojo-1000 introduced some changes yesterday, which have already been rebased into the development branch, did you already take a look at these?

**Comment by herbrechtsmeier** [comment]
@enwi I have take a look at the changes. Maybe the groups handling could move into its own class to separate it from the hue class.

Why is the class called `Group` and not `HueGroup`? Should I remove the `Hue` from the `Sensor` class?

Why you cache the requests and not the system state? A request to `/api/<apikey>/`, `/api/<apikey>/lights` and `/api/<apikey>/lights/<id>` share the same state of a light.

**Comment by Jojo-1000** [comment]
@herbrechtsmeier The Hue in HueLight is from the time when we had no namespace. I found it redundant to have hueplusplus::HueGroup instead of hueplusplus::Group.

I am not sure if sharing so much state between the classes is a good idea. It makes it difficult to see what is going on from an outside perspective, and it prevents use in a multithreaded environment. I would maybe even remove the map of cached HueLights from Hue and instead have the user handle that if they want to.

Regarding the tests, feel free to rewrite them without mocks, but I believe that will make them less maintainable.

**Comment by herbrechtsmeier** [comment]
> @herbrechtsmeier The Hue in HueLight is from the time when we had no namespace. I found it redundant to have hueplusplus::HueGroup instead of hueplusplus::Group.

Should I remove the `Hue` from `HueSensor`?

> I am not sure if sharing so much state between the classes is a good idea. It makes it difficult to see what is going on from an outside perspective, and it prevents use in a multithreaded environment.

Agree

> I would maybe even remove the map of cached HueLights from Hue and instead have the user handle that if they want to.

Does this means `getAllLights()` would request `/api/<apikey>/lights` and creates a vector of objects on the fly and `getLight()` creates a new object?

**Comment by Jojo-1000** [comment]
> Does this means `getAllLights()` would request `/api/<apikey>/lights` and creates a vector of objects on the fly and `getLight()` creates a new object?

Yes, or you could use the current state of the bridge to create the new objects.

The way it currently is, there should only be one instance of every light, but the user can just do
`HueLight light = bridge.getLight(1)` and you have an unnecessary copy sitting in the bridge cache.

**Comment by enwi** [comment]
> Should I remove the `Hue` from `HueSensor`?

Also to not let your question be unanswered, I would say yes

**Comment by Jojo-1000** [comment]
@herbrechtsmeier I am now at the point where the sensor and rule apis are pretty much the only missing pieces, so I could start working on sensor support. Would you like to continue working on this issue yourself, or should I take over here?

**Comment by herbrechtsmeier** [comment]
@Jojo-1000 Please take over. The HUE sensors and especially the CLIP sensors are much more heterogeneous as expected.

**Pull request merged**